Core/ExpectedException/ExpectedExceptionModel.cs

```diff
@@ -0,0 +1,179 @@
+// and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// 
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+#endregion
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using NUnit2To3SyntaxConverter.Extensions;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+using static NUnit2To3SyntaxConverter.Extensions.SyntaxFactoryUtils;
+
+namespace NUnit2To3SyntaxConverter.ExpectedException
+{
+  public class ExpectedExceptionModel : IExpectedExceptionModel
+  {
+    private readonly AttributeData _attributeData;
+    public ExpressionSyntax ExceptionType { get; }
+    public ExpressionSyntax? UserMessage { get; }
+    public ExpressionSyntax? MatchType { get; }
+    public ExpressionSyntax? ExpectedMessage { get; }
+
+    public ExpectedExceptionModel (
+        AttributeData attributeData,
+        ExpressionSyntax exceptionType,
+        ExpressionSyntax? userMessage,
+        ExpressionSyntax? expectedMessage,
+        ExpressionSyntax? matchType)
+    {
+      _attributeData = attributeData;
+      ExceptionType = exceptionType;
+      UserMessage = userMessage;
+      MatchType = matchType;
+      ExpectedMessage = expectedMessage;
+    }
+
+    public async Task<AttributeSyntax> GetAttributeSyntax ()
+    {
+      return await _attributeData.ApplicationSyntaxReference.GetSyntaxAsync() as AttributeSyntax
+             ?? throw new InvalidOperationException ($"member {_attributeData} is not a AttributeSyntax.");
+    }
+
+    public ExpressionSyntax AsConstraintExpression (string baseIndent)
+    {
+      var exception = (ExceptionType as TypeOfExpressionSyntax)
+                      ?.Type.ToString()
+                      ?? nameof(Exception);
+
+      var throwsWithInstanceOfExpressionType = ExceptionConstraintSyntaxFactory.CreateThrowsExceptionConstrainSyntax (exception)
+          .WithLeadingTrivia (Whitespace (baseIndent));
+
+      return ExpectedMessage == null
+          ? throwsWithInstanceOfExpressionType
+          : CreateWithMessage (throwsWithInstanceOfExpressionType, baseIndent);
+    }
+
+    public static ExpectedExceptionModel CreateFromAttributeData (AttributeData attribute)

```

Comment:
public (or private) static should be placed before the type's state. private static should normally only be static if it's required for a public static method call. If a private's static in order to prevent capturing the THIS pointer etc, it can remain with the instance methods.